experiment/autobumper/bumper/bumper.go

```diff
@@ -68,49 +247,145 @@ func (w HideSecretsWriter) Write(content []byte) (int, error) {
+		if err != nil {
+			return nil, fmt.Errorf("failed to resolve the upstream staging Prow image version: %w", err)
+		}
+	default:
+		tagPicker = func(imageHost, imageName, currentTag string) (string, error) { return tagVersion, nil }
+	}
+
+	updateFile := func(name string) error {
+		logrus.Infof("Updating file %s", name)
+		if err := imagebumper.UpdateFile(tagPicker, name, filterRegexp); err != nil {
+			return fmt.Errorf("failed to update the file: %w", err)
+		}
+		return nil
+	}
+	updateYAMLFile := func(name string) error {
+		if strings.HasSuffix(name, ".yaml") && !isUnderPath(name, excludeConfigPaths) {
+			return updateFile(name)
+		}
+		return nil
+	}
+
+	// Updated all .yaml files under the included config paths but not under excluded config paths.
+	for _, path := range includeConfigPaths {
+		info, err := os.Stat(path)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get the file info for %q", path)
+		}
+		if info.IsDir() {
+			err := filepath.Walk(path, func(subpath string, info os.FileInfo, err error) error {
+				return updateYAMLFile(subpath)
+			})
+			if err != nil {
+				return nil, fmt.Errorf("failed to update yaml files under %q: %w", path, err)
+			}
+		} else {
+			if err := updateYAMLFile(path); err != nil {
+				return nil, fmt.Errorf("failed to update the yaml file %q: %w", path, err)
+			}
+		}
+	}
+
+	// Update the extra files in any case.
+	for _, file := range extraFiles {
+		if err := updateFile(file); err != nil {
+			return nil, fmt.Errorf("failed to update the extra file %q: %w", file, err)
 		}
 	}
 
-	return bumper.GetReplacements(), nil
+	return imagebumper.GetReplacements(), nil
+}
+
+func upstreamImageVersionResolver(upstreamAddress string) (func(imageHost, imageName, currentTag string) (string, error), error) {
+	version, err := parseUpstreamImageVersion(upstreamAddress)
+	if err != nil {
+		return nil, fmt.Errorf("error resolving the upstream Prow version from %q: %w", upstreamAddress, err)
+	}
+	return func(imageHost, imageName, currentTag string) (string, error) {
+		// Skip boskos images as they do not have the same image tag as other Prow components.
+		// TODO(chizhg): remove this check after all Prow instances are using boskos images not in gcr.io/k8s-prow/boskos
+		if strings.Contains(imageName, "boskos/") {
+			return currentTag, nil
+		}
+		return version, nil
+	}, nil
+}
+
+func parseUpstreamImageVersion(upstreamAddress string) (string, error) {
+	resp, err := http.Get(upstreamAddress)
+	if err != nil {
+		return "", fmt.Errorf("error sending GET request to %q: %w", upstreamAddress, err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("HTTP error %d (%q) fetching upstream config file", resp.StatusCode, resp.Status)
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("error reading the response body: %w", err)
+	}
+	imagematcher := regexp.MustCompile(`(?s)^.+image:.+:(v[a-zA-Z0-9_.-]+)`)
```

Comment:
move this into a package-level var so we don't initialize it every call (and/or return the error)

Reply:
Done.